src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -0,0 +1,369 @@
+﻿using System.Buffers;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO.Pipelines
+{
+    internal class StreamPipeReader : PipeReader
+    {
+        internal const int InitialSegmentPoolSize = 4; // 16K
+        internal const int MaxSegmentPoolSize = 256; // 1MB
+
+        private readonly int _bufferSize;
+        private readonly int _minimumReadThreshold;
+        private readonly MemoryPool<byte> _pool;
+
+        private CancellationTokenSource _internalTokenSource;
+        private bool _isReaderCompleted;
+        private bool _isStreamCompleted;
+
+        private BufferSegment _readHead;
+        private int _readIndex;
+
+        private BufferSegment _readTail;
+        private long _bufferedBytes;
+        private bool _examinedEverything;
+        private object _lock = new object();
+
+        // Mutable struct! Don't make this readonly
+        private BufferSegmentStack _bufferSegmentPool;
+
+        /// <summary>
+        /// Creates a new StreamPipeReader.
+        /// </summary>
+        /// <param name="readingStream">The stream to read from.</param>
+        /// <param name="options">The options to use.</param>
+        public StreamPipeReader(Stream readingStream, StreamPipeReaderOptions options)
+        {
+            InnerStream = readingStream ?? throw new ArgumentNullException(nameof(readingStream));
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            if (options.MinimumReadSize <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(options.MinimumReadSize));
+            }
+
+            _bufferSegmentPool = new BufferSegmentStack(InitialSegmentPoolSize);
+            _minimumReadThreshold = Math.Min(options.MinimumReadSize, options.BufferSize);
+            _pool = options.Pool == MemoryPool<byte>.Shared ? null : options.Pool;
+            _bufferSize = _pool == null ? options.BufferSize : Math.Min(options.BufferSize, _pool.MaxBufferSize);
+        }
+
+        /// <summary>
+        /// Gets the inner stream that is being read from.
+        /// </summary>
+        public Stream InnerStream { get; }
+
+        /// <inheritdoc />
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            AdvanceTo(consumed, consumed);
+        }
+
+        private CancellationTokenSource InternalTokenSource
+        {
+            get
+            {
+                lock (_lock)
+                {
+                    if (_internalTokenSource == null)
+                    {
+                        _internalTokenSource = new CancellationTokenSource();
+                    }
+                    return _internalTokenSource;
+                }
+            }
+            set
+            {
+                _internalTokenSource = value;
```

Comment:
Any reason this assignment is no syncronized?

Reply:
I think you can just remove the setter. It isn't used.